src/graph/sampling/randomwalks/get_node_types_gpu.cu

```diff
@@ -0,0 +1,90 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/sampling/get_node_types_gpu.cu
+ * \brief DGL sampler
+ */
+
+#include <dgl/array.h>
+#include <dgl/base_heterograph.h>
+#include <dgl/runtime/device_api.h>
+#include <cuda_runtime.h>
+#include <utility>
+#include "randomwalks_impl.h"
+
+namespace dgl {
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace sampling {
+
+namespace impl {
+
+template<DLDeviceType XPU, typename IdxType>
+TypeArray GetNodeTypesFromMetapath(
+    const HeteroGraphPtr hg,
+    const TypeArray metapath) {
+
+  uint64_t num_etypes = metapath->shape[0];
+  TypeArray result = TypeArray::Empty(
+      {metapath->shape[0] + 1}, metapath->dtype, metapath->ctx);
+  IdxType *result_data = static_cast<IdxType *>(result->data);
```

Comment:
`static_cast<T *>(array->data)` could be replaced with `array.Ptr<T>()` which is more concise.